CMakeLists.txt

```diff
@@ -107,6 +107,17 @@ if (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "Clang")
   endif ()
 endif ()
 
+# when cross-compiling, tell GCC linker to use static libraries if possible
+# in order to avoid dependencies on libgcc and libstdc++ shared libraries
+# (this produces self-contained but larger binaries)
+# (NOTE: pay attention if C++ exceptions are propagated across libraries)
+if (CMAKE_CROSSCOMPILING AND CMAKE_COMPILER_IS_GNUCC AND BUILD_SHARED_LIBS)
+  set (CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -static")
+  set (CMAKE_STATIC_LINKER_FLAGS "${CMAKE_STATIC_LINKER_FLAGS} -static")
+  set (CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -static")
+  set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static")
+endif ()
+
```

Comment:
I guess we could add an `option()` to control this ON or OFF when cross-compiling..

But just to be clear, the `-static` link flag does not mess with `BUILD_SHARED_LIBS` setting, a shared library (DLL) will still be generated if it was requested (i.e `-static` and `-shared` GCC flags can be used together). It is only telling the linker not to depend on MinGW DLLs same as using the flags `-static-libgcc` and `-static-libstdc++`:

> https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#index-static-1164

To illustrate, without using `-static` flag we get this:

``` sh
# `objdump` is used here as a poor substitute of `ldd` for windows binaries

# WITH_CXX=OFF BUILD_SHARED_LIBS=ON
$ x86_64-w64-mingw32-objdump -p libnlopt.dll | grep 'DLL Name'
    DLL Name: KERNEL32.dll
    DLL Name: msvcrt.dll
    DLL Name: libgcc_s_sjlj-1.dll

# WITH_CXX=ON BUILD_SHARED_LIBS=ON
$ x86_64-w64-mingw32-objdump -p libnlopt_cxx.dll | grep 'DLL Name'
    DLL Name: KERNEL32.dll
    DLL Name: msvcrt.dll
    DLL Name: libgcc_s_sjlj-1.dll
    DLL Name: libstdc++-6.dll
```

Using `-static` we get this:

``` sh
# WITH_CXX=OFF BUILD_SHARED_LIBS=ON
$ x86_64-w64-mingw32-objdump -p libnlopt.dll | grep 'DLL Name'
    DLL Name: KERNEL32.dll
    DLL Name: msvcrt.dll

# WITH_CXX=ON BUILD_SHARED_LIBS=ON
$ x86_64-w64-mingw32-objdump -p libnlopt_cxx.dll | grep 'DLL Name'
    DLL Name: KERNEL32.dll
    DLL Name: msvcrt.dll
```

So without `-static` those extra libgcc and libstdc++ dependencies will have to be copied as well when deploying cross-compiled binaries.

Of course the downside is that the binaries are larger.